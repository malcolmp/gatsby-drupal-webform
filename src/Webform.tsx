import React, { useState, FormEvent } from 'react'
import axios from 'axios'
import './fragments'

import { deNormalizeElement, formToJSON } from './utils'
import { WebformInput, WebformSelect, WebformTextarea, WebformCheckbox, WebformCheckboxGroup, WebformText } from './components'

export const DEFAULT_SUBMIT_LABEL = 'Submit'

/**
 * Webform object as returned from GraphQL query.
 */
export interface WebformObject {
	drupal_internal__id: string
	description: string
	status: string
	elements: WebformElement[]
}

/**
 * Webform element (e.g. text field input or submit button) as returned by GraphQL.
 */
export type WebformElement = {
	name: string
	title: string
	type: string
	attributes?: WebformElementAttribute[]
	label_attributes?: WebformElementAttribute[]
	wrapper_attributes?: WebformElementAttribute[]
	options?: Array<{
		label: string
		value: string
	}>
	states?: Array<{
		state: string
		selector: string
		condition: {
			[key: string]: boolean | string | null
		}
	}>
	additional_properties?: WebformElementAttribute[]
}

/**
 * A name value mapping of element attributes.
 */
export type WebformElementAttribute = {
	name: string
	value: string
}

/**
 * A de-normalized webform element.
 */
export type DeNormalizedWebformElement = {
	name: string
	title: string
	type: string
	attributes: {
		[key: string]: any
	}
	label_attributes: {
		[key: string]: any
	}
	wrapper_attributes: {
		[key: string]: any
	}
	options?: Array<{
		label: string
		value: string
	}>
	states?: {
		[key: string]: boolean
	}
	additional_properties: {
		[key: string]: any
	}
}

export type WebformComponentProps = {
	element: WebformElement
	error?: string
}

/**
 * Custom component for webform element
 */
export type WebformComponent = React.FC<WebformComponentProps>

export type WebformValidateHandler = (event: FormEvent<HTMLFormElement>) => boolean
export type WebformSubmitHandler = (
	data: ReturnType<typeof formToJSON>,
	event: FormEvent<HTMLFormElement>
) => void | boolean | Promise<void | boolean>

export type WebformSuccessHandler = (response: any, event: FormEvent<HTMLFormElement>) => void
export type WebformErrorHandler = (err: any, event: FormEvent<HTMLFormElement>) => void

export class WebformError extends Error {
	response: any

	constructor(response: any) {
		super()

		this.response = response
	}
}

/**
 * Errors returned by Drupal.
 */
type WebformErrors = {
	[name: string]: string
}

interface Props {
	id?: string
	className?: string
	style?: React.CSSProperties
	noValidate?: boolean

	webform: WebformObject

	onValidate?: WebformValidateHandler
	onSuccess?: WebformSuccessHandler
	onError?: WebformErrorHandler

	/**
	 * Called right before webform data is POSTed to the api.
	 *
	 * If callback returns false nothing is sent to the api.
	 */
	onSubmit?: WebformSubmitHandler

	/** Form POST endpoint. */
	endpoint: string

	/** Extra data to POST. */
	extraData?: object

	/** Provide custom components that handle specific webform elements. */
	customComponents: { [name: string]: WebformComponent }
}

/**
 * Render single webform element.
 */
export function renderWebformElement(element: WebformElement, error?: string, CustomComponent?: WebformComponent) {
	const customComponentAPI = {
		error
	}

	// Render using custom component if provided:
	if (CustomComponent) {
		return <CustomComponent element={element} {...customComponentAPI} />
	}

	// Otherwise select renderer based on element type:
	switch (element.type) {
		case 'textfield':
			return <WebformInput element={{ ...element, type: 'text' }} {...customComponentAPI} />
		case 'textarea':
			return <WebformTextarea element={element} {...customComponentAPI} />
		case 'tel':
		case 'number':
		case 'email':
		case 'hidden':
			return <WebformInput element={element} {...customComponentAPI} />
		case 'checkbox':
		case 'radio':
			/** Render single checkbox or radio element. */
			return <WebformCheckbox element={element} {...customComponentAPI} />
		case 'checkboxes':
			return <WebformCheckboxGroup element={{ ...element, type: 'checkbox' }} {...customComponentAPI} />
		case 'radios':
			return <WebformCheckboxGroup element={{ ...element, type: 'radio' }} {...customComponentAPI} />
		case 'select':
			return <WebformSelect element={element} {...customComponentAPI} />
		case 'processed_text':
			return <WebformText element={element} {...customComponentAPI} />
		// Submit button
		case 'webform_actions':
			const el = deNormalizeElement(element)
			const submitLabel = (el.additional_properties || {}).submit__label || DEFAULT_SUBMIT_LABEL
			return (
				<div className="form-group">
					<button type="submit">{submitLabel}</button>
				</div>
			)
		// Unknown element type -> render an empty string.
		default:
			return ''
	}
}

/**
 * Drupal webform react component.
 */
const Webform = ({ webform, customComponents, ...props }: Props) => {
	const [errors, setErrors] = useState<WebformErrors>({})

	const submitHandler: React.FormEventHandler<HTMLFormElement> = async event => {
		event.preventDefault()

		const target = event.currentTarget

		// Clear errors from previous submit.
		setErrors({})

		// Remove lingering css classes from previous submits.
		target.classList.remove('form-submitting', 'form-error', 'form-submitted')

		if ((!props.onValidate || props.onValidate(event)) && target.checkValidity()) {
			// Let css know that this form was validated and is being submitted.
			target.classList.add('was-validated', 'form-submitting')

			// Serialize form data.
			const data = formToJSON(target.elements)

			// Post process serialized data:
			// Some webform elements require specialized data formatting.
			for (const element of webform.elements) {
				if (data[element.name]) {
					switch (element.type) {
						case 'checkbox':
							data[element.name] = data[element.name][0]
							break
					}
				}
			}

			try {
				// If onSubmit returns false skip submitting to API.
				if (props.onSubmit && (await props.onSubmit(data, event)) === false) {
					target.classList.replace('form-submitting', 'form-submitted')
					return
				}

				// Submit form to API.
				const response = await axios.post(props.endpoint, {
					...props.extraData,
					...data,
					webform_id: webform.drupal_internal__id
				})

				if (response.data.error) {
					throw new WebformError(response)
				}

				// Convey current form state.
				target.classList.replace('form-submitting', 'form-submitted')
				props.onSuccess && props.onSuccess(response.data, event)
			} catch (err) {
				// API should return error structure if validation fails.
				// We use that to render error messages to the form.
				if (err.response && err.response.data.error) {
					setErrors(err.response.data.error)
				}

				// Convey current form state.
				target.classList.replace('form-submitting', 'form-error')
				props.onError && props.onError(err, event)
			}
		} else {
			// Let css know this form was validated.
			target.classList.add('was-validated', 'form-error')
		}
	}

	return (
		<form
			onSubmit={submitHandler}
			id={props.id}
			className={props.className}
			style={props.style}
			noValidate={props.noValidate}
			data-webform-id={webform.drupal_internal__id}
		>
			{/* Render webform elements */}
			{webform.elements.map(element => (
				<React.Fragment key={element.name}>
					{renderWebformElement(element, errors[element.name], customComponents![element.type])}
				</React.Fragment>
			))}

			{/* Render default submit button if it is not defined in elements array. */}
			{webform.elements.find(element => element.type === 'webform_actions') === undefined && (
				<button type="submit">{DEFAULT_SUBMIT_LABEL}</button>
			)}
		</form>
	)
}

Webform.defaultProps = {
	customComponents: {}
}

export default Webform
