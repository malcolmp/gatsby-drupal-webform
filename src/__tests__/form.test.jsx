import React from 'react'
import { render } from '@testing-library/react'

import Webform from '../Webform'

const webform = {
	drupal_internal__id: 'contact',
	elements: [
		{
			name: 'name',
			title: 'Your Name',
			type: 'textfield',
			attributes: [
				{
					name: 'required',
					value: 'true'
				},
				{
					name: 'defaultValue',
					value: 'John'
				}
			]
		},
		{
			name: 'email',
			title: 'Your Email',
			type: 'email',
			attributes: [
				{
					name: 'required',
					value: 'true'
				},
				{
					name: 'defaultValue',
					value: 'john@example.com'
				},
				{
					name: 'autocomplete',
					value: 'email'
				}
			]
		},
		{
			name: 'subject',
			title: 'Subject',
			type: 'email',
			attributes: [
				{
					name: 'required',
					value: 'true'
				}
			]
		},
		{
			name: 'message',
			title: 'Subject',
			type: 'textarea',
			attributes: [
				{
					name: 'required',
					value: 'true'
				}
			]
		},
		{
			name: 'actions',
			title: 'Submit button(s)',
			type: 'webform_actions',
			additional_properties: [
				{
					name: 'submit__label',
					value: 'Send message'
				}
			]
		}
	]
}

describe('Webform', () => {
	it('contact form renders correctly', () => {
		const { asFragment } = render(<Webform className="my-form" style={{ color: 'black' }} endpoint="/form-submit" webform={webform} />)

		expect(asFragment()).toMatchSnapshot()
	})

	it('title attributes', () => {
		const webform = {
			drupal_internal__id: 'contact',
			elements: [
				{
					name: 'name',
					title: 'Your Name',
					type: 'textfield',
					additional_properties: [
						{
							name: 'title_display',
							value: 'invisible'
						}
					]
				}
			]
		}

		const { asFragment } = render(<Webform className="my-form" endpoint="/form-submit" webform={webform} />)

		expect(asFragment()).toMatchSnapshot()
	})

	it('hidden inputs', () => {
		const webform = {
			drupal_internal__id: 'contact',
			elements: [
				{
					name: 'name',
					title: 'Your Name',
					type: 'hidden',
				}
			]
		}

		const { asFragment } = render(<Webform className="my-form" endpoint="/form-submit" webform={webform} />)
		expect(asFragment()).toMatchSnapshot()
	})

	it('checkboxes', () => {
		const webform = {
			drupal_internal__id: 'contact',
			elements: [
				{
					name: 'tos',
					title: 'Agree terms of service',
					type: 'checkbox',
				},
				{
					name: 'favourite_color',
					title: 'Select your favourite color',
					type: 'radios',
					options: [
						{
							label: 'Hotpink',
							value: '#ff69b4'
						},
						{
							label: 'Olive',
							value: '#808000'
						},
						{
							label: 'Cornflower blue',
							value: '#6495ed'
						}
					]
				}
			]
		}

		const { asFragment } = render(<Webform className="my-form" endpoint="/form-submit" webform={webform} />)

		expect(asFragment()).toMatchSnapshot()
	})
})
