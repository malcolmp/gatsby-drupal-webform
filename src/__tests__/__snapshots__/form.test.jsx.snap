// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Webform checkboxes 1`] = `
<DocumentFragment>
  <form
    class="my-form"
    data-webform-id="contact"
  >
    <div
      class="form-group form-check"
    >
      <input
        class="form-check-input"
        id="form-tos"
        name="tos"
        type="checkbox"
        value="1"
      />
      <label
        class="form-check-label"
        for="form-tos"
      >
        Agree terms of service
      </label>
    </div>
    <div
      class="form-group"
    >
      <label>
        Select your favourite color
      </label>
      <div
        class="form-check"
      >
        <input
          class="form-check-input"
          id="form-favourite-color-#ff69b4"
          name="favourite_color"
          type="radio"
          value="#ff69b4"
        />
        <label
          class="form-check-label"
          for="form-favourite-color-#ff69b4"
        >
          Hotpink
        </label>
      </div>
      <div
        class="form-check"
      >
        <input
          class="form-check-input"
          id="form-favourite-color-#808000"
          name="favourite_color"
          type="radio"
          value="#808000"
        />
        <label
          class="form-check-label"
          for="form-favourite-color-#808000"
        >
          Olive
        </label>
      </div>
      <div
        class="form-check"
      >
        <input
          class="form-check-input"
          id="form-favourite-color-#6495ed"
          name="favourite_color"
          type="radio"
          value="#6495ed"
        />
        <label
          class="form-check-label"
          for="form-favourite-color-#6495ed"
        >
          Cornflower blue
        </label>
      </div>
    </div>
    <button
      type="submit"
    >
      Submit
    </button>
  </form>
</DocumentFragment>
`;

exports[`Webform contact form renders correctly 1`] = `
<DocumentFragment>
  <form
    class="my-form"
    data-webform-id="contact"
    style="color: black;"
  >
    <div
      class="form-group required"
    >
      <label
        for="form-name"
      >
        Your Name
      </label>
      <input
        class="form-control"
        id="form-name"
        name="name"
        required=""
        type="text"
        value="John"
      />
    </div>
    <div
      class="form-group required"
    >
      <label
        for="form-email"
      >
        Your Email
      </label>
      <input
        autocomplete="email"
        class="form-control"
        id="form-email"
        name="email"
        required=""
        type="email"
        value="john@example.com"
      />
    </div>
    <div
      class="form-group required"
    >
      <label
        for="form-subject"
      >
        Subject
      </label>
      <input
        class="form-control"
        id="form-subject"
        name="subject"
        required=""
        type="email"
      />
    </div>
    <div
      class="form-group required"
    >
      <label
        for="form-message"
      >
        Subject
      </label>
      <textarea
        class="form-control"
        id="form-message"
        name="message"
        required=""
      />
    </div>
    <div
      class="form-group"
    >
      <button
        type="submit"
      >
        Send message
      </button>
    </div>
  </form>
</DocumentFragment>
`;

exports[`Webform hidden inputs 1`] = `
<DocumentFragment>
  <form
    class="my-form"
    data-webform-id="contact"
  >
    <input
      class="form-control"
      id="form-name"
      name="name"
      type="hidden"
    />
    <button
      type="submit"
    >
      Submit
    </button>
  </form>
</DocumentFragment>
`;

exports[`Webform title attributes 1`] = `
<DocumentFragment>
  <form
    class="my-form"
    data-webform-id="contact"
  >
    <div
      class="form-group"
    >
      <label
        for="form-name"
        style="position: absolute; overflow: hidden; clip: rect(1px, 1px, 1px, 1px); width: 1px; height: 1px; word-wrap: normal;"
      >
        Your Name
      </label>
      <input
        class="form-control"
        id="form-name"
        name="name"
        type="text"
      />
    </div>
    <button
      type="submit"
    >
      Submit
    </button>
  </form>
</DocumentFragment>
`;
